### PR TITLE
Fix config file search locations (should fix #1538)

### LINKED ISSUE #1534: Test failures with Python 3.10.0
I get four test failures:

FAILED psyad/main_test.py::test_main_h_option - AssertionError: assert '[-h] ...
FAILED psyad/transformations/test_adjoint_trans.py::test_abstract - assert "C...
FAILED psyir/symbols/datatype_test.py::test_datatype - assert "Can't instanti...
FAILED psyir/transformations/intrinsics/operator2code_trans_test.py::test_create

These are due to text differences in the usage message and in the error reported when an attempt is made to instantiate an abstract class.

### LINKED ISSUE #1538: RTD UG build failing to include autodoc documentation for transformations
Comparing a local build of the UG with that on RTD, it is clear that the documentation for a lot of the transformations is missing. On looking at the logs I see:

    psyclone.configuration.ConfigurationError: PSyclone configuration error: psyclone.cfg not found in any of ['/home/docs/checkouts/readthedocs.org/user_builds/psyclone/checkouts/latest/doc/user_guide/.psyclone', '/home/docs/checkouts/readthedocs.org/user_builds/psyclone/envs/latest/share/psyclone', '/home/docs/.local/share/psyclone']

    WARNING: autodoc: failed to import class 'lfric.transformations.LFRicExtractTrans' from module 'psyclone.domain'; the following exception was raised:
    Traceback (most recent call last):
      File "/home/docs/checkouts/readthedocs.org/user_builds/psyclone/envs/latest/lib/python3.7/site-packages/sphinx/ext/autodoc/importer.py", line 70, in import_module
    return importlib.import_module(modname)
      File "/home/docs/checkouts/readthedocs.org/user_builds/psyclone/envs/latest/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)

### PR DESCRIPTION
**Pull request by nmnobre**
Now that I've spent some time fixing docstrings, I'd rather like to see them show up in the documentation!

### CODE REVIEW

**Review by arporter — CHANGES_REQUESTED** [review]
Thanks for taking this on Nuno.
I've setup a RTD build for this branch (https://psyclone.readthedocs.io/en/fix_find_config/) and things are still not working but for new and exciting reasons:

    WARNING: autodoc: failed to import module 'generator' from module 'psyclone'; the following exception was raised:
    Python version 3.8 or above is required for SymPy.
    WARNING: don't know which module to import for autodocumenting 'generate' (try placing a "module" or "currentmodule" directive in the document, or giving an explicit module name)
    WARNING: autodoc: failed to import module 'transformations' from module 'psyclone'; the following exception was raised:
    Python version 3.8 or above is required for SymPy.

I've done a bit of googling but to no avail. I don't understand the error message because we run the test suite for Python 3.6 quite happily. (It seems that RTD is using 3.7.) If you have any ideas, that would be great. If not, I'll still take this addition as it certainly is a step in the right direction.

**Comment by arporter** [comment]
An update: I've installed Python 3.7.14 locally and the user guide builds fine so I really don't understand what's wrong on RTD.

**Comment by arporter** [comment]
We can tell RTD to use `pip` but it requires a `.readthedocs.yaml` file. Previously I've stayed away from that as we're using RTD to build three different docs from a single repo (and thus we have three different conf files). Can we use the yaml file but not specify the conf file in it?

**Comment by arporter** [comment]
With the new yaml file, both the Dev Guide and User Guide build correctly for this branch! So, if you could do the requested tidying and then request a review from someone other than me that would be great :-)

**Comment by nmnobre** [comment]
> Thanks for taking this on Nuno. I've setup a RTD build for this branch (https://psyclone.readthedocs.io/en/fix_find_config/) and things are still not working but for new and exciting reasons:
> 
> ```
> WARNING: autodoc: failed to import module 'generator' from module 'psyclone'; the following exception was raised:
> Python version 3.8 or above is required for SymPy.
> WARNING: don't know which module to import for autodocumenting 'generate' (try placing a "module" or "currentmodule" directive in the document, or giving an explicit module name)
> WARNING: autodoc: failed to import module 'transformations' from module 'psyclone'; the following exception was raised:
> Python version 3.8 or above is required for SymPy.
> ```
> 
> I've done a bit of googling but to no avail. I don't understand the error message because we run the test suite for Python 3.6 quite happily. (It seems that RTD is using 3.7.) If you have any ideas, that would be great. If not, I'll still take this addition as it certainly is a step in the right direction.

Well, I know this is _not_ new. This error was already happening for builds of version latest. I didn't give it much thought because it wasn't happening for builds of version stable and so I deemed it unrelated to the missing transformation docs.

**Comment by nmnobre** [comment]
> An update: I've installed Python 3.7.14 locally and the user guide builds fine so I really don't understand what's wrong on RTD.
> It turns out that that's true so long as I use `pip` when installing PSyclone. If I do what RTD does and use `python setup.py install` then I _do_ see the same errors when building the UG.

Exactly. The problem is the installation location used by each method.

**Comment by nmnobre** [comment]
Ok - I think understand it now, here's what my investigation found:

1) As discussed before, `pip` and `setup.py install` unfortunately don't place `psyclone.cfg` at the same location. My additions fix this.

2) A second issue has to do with SymPy. This is happening since readthedocs decided to update from SymPy 1.10.1 to 1.11rc1, somewhere between 08/08/22 and 11/08/22. Here's a build from the 8th, https://readthedocs.org/projects/psyclone/builds/17652118/, and a second build from the 11th, https://readthedocs.org/projects/psyclone/builds/17685498/. SymPy has since been updated to 1.11.1 and still requires Python >=3.8. Now, the question is, why did readthedocs bump SymPy but not Python? Are we (unintentionally perhaps?) blocking updates to Python? There's no reason to not always use the latest and greatest to build the docs.

3) Given the above, either we change to use the yaml file you created which switches to both `pip` and to Python version 3.8, solving both issues or we keep my fix and find a way to tell readthedocs to not use Python 3.7, is this possible?

Cheers :-)

**Comment by arporter** [comment]
> 3\. Given the above, either we change to use the yaml file you created which switches to both `pip` and to Python version 3.8, solving both issues or we keep my fix and find a way to tell readthedocs to not use Python 3.7, is this possible?

By not using the yaml file we may have been unintentionally blocking things. All of RTD's docs say to use a configuration file and I saw somewhere that the `python setup.py install` is deprecated. I vote we keep the yaml *and* your fixes.

**Comment by nmnobre** [comment]
> > 3. Given the above, either we change to use the yaml file you created which switches to both `pip` and to Python version 3.8, solving both issues or we keep my fix and find a way to tell readthedocs to not use Python 3.7, is this possible?
> 
> By not using the yaml file we may have been unintentionally blocking things. All of RTD's docs say to use a configuration file and I saw somewhere that the `python setup.py install` is deprecated. I vote we keep the yaml _and_ your fixes.

The docs say "In addition to using the admin panel of your project to configure your project, you can use a configuration file in the root of your project.", "In addition" being the key phrase here?

File: src/psyclone/configuration.py [hunk 1]
```diff
@@ -427,6 +430,8 @@ def find_file():
 
         # Set up list of locations to search
         share_dir = os.path.join(sys.prefix, "share", "psyclone")
+        pkg_share_dir = [os.path.join(psyclone_path, "..", "share", "psyclone")
```
[the remark this answers is not in the document]
Comment (nmnobre) [reply src/psyclone/configuration.py:433]: Done.

File: src/psyclone/configuration.py [hunk 2]
```diff
@@ -439,6 +444,8 @@ def find_file():
         if not within_virtual_env():
             # 4. <python-installation-base>/share/psyclone/
             _file_paths.append(share_dir)
+        # 5. <psyclone-installation-base>/share/psyclone/
```
[the remark this answers is not in the document]
Comment (nmnobre) [reply src/psyclone/configuration.py:447]: Done?

**Review by arporter — APPROVED** [review]
All good now. My test build on RTD is still working fine.
Changes are pylint clean (apart from one tiny thing that I've taken the liberty of fixing).
Tests and examples are all OK with compilation.
#1538 isn't mentioned anywhere.
Will proceed to merge.

**Comment by arporter** [comment]
Link test failure is graphviz temporary failure (link works OK when I test it).

**Pull request merged**